<a href="https://colab.research.google.com/github/johnjdailey/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/John_Dailey_LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

1. Load and clean the data (or determine the best method to drop observations when running tests)

In [0]:
### YOUR CODE STARTS HERE ###
from scipy.stats import ttest_ind, ttest_ind_from_stats, ttest_rel

In [0]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-01-30 02:56:36--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data.4’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.06s   

2020-01-30 02:56:37 (278 KB/s) - ‘house-votes-84.data.4’ saved [18171/18171]



In [0]:
import pandas as pd
import numpy as np

In [0]:
column_headers = ['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa']

df = pd.read_csv('house-votes-84.data',
                 header=None,
                 names=column_headers,
                 na_values='?')

df.head()

party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n          NaN
2    democrat                 NaN             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [0]:
df = df.replace({'y': 1, 'n': 0})
df.head

<bound method NDFrame.head of           party  handicapped-infants  ...  duty-free  south-africa
0    republican                  0.0  ...        0.0           1.0
1    republican                  0.0  ...        0.0           NaN
2      democrat                  NaN  ...        0.0           0.0
3      democrat                  0.0  ...        0.0           1.0
4      democrat                  1.0  ...        1.0           1.0
..          ...                  ...  ...        ...           ...
430  republican                  0.0  ...        0.0           1.0
431    democrat                  0.0  ...        0.0           1.0
432  republican                  0.0  ...        0.0           1.0
433  republican                  0.0  ...        0.0           1.0
434  republican                  0.0  ...        NaN           0.0

[435 rows x 17 columns]>

2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01

In [0]:
rep = df[df['party']=='republican']
rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [0]:
col = rep['aid-to-contras']

np.isnan(col)

aid_to_contras_no_nans = col[~np.isnan(col)]

aid_to_contras_no_nans

aid_to_contras_no_nans.sum()/len(aid_to_contras_no_nans)

0.15286624203821655

In [0]:
rep['aid-to-contras'].mean()

0.15286624203821655

In [0]:
dem = df[df['party']=='democrat']
dem.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

In [0]:
col = dem['aid-to-contras']

np.isnan(col)

aid_to_contras = col[~np.isnan(col)]

aid_to_contras_no_nans

aid_to_contras_no_nans.sum()/len(aid_to_contras_no_nans)

0.15286624203821655

In [0]:
dem['aid-to-contras'].mean()

0.8288973384030418

In [0]:
ttest_ind(rep['aid-to-contras'], dem['aid-to-contras'])

Ttest_indResult(statistic=nan, pvalue=nan)

In [0]:
ttest_ind(rep['aid-to-contras'], dem['aid-to-contras'], nan_policy='omit')

Ttest_indResult(statistic=-18.052093200819733, pvalue=2.82471841372357e-54)

In [0]:
col = rep['aid-to-contras']
rep_aid_to_contras_no_nans = col[~np.isnan(col)]

col = dem['aid-to-contras']
dem_aid_to_contras_no_nans = col[~np.isnan(col)]

print(len(rep_aid_to_contras_no_nans))
print(len(dem_aid_to_contras_no_nans))

157
263


df = 157-1 = 156

T-statistic: -18.052093200819733

P-value: 2.82471841372357e-54

I want to reject the null hypothesis if my p-value is <.01 or if my p-value is less than (1-confidence_level)

Conclusion: due to a p-value of 2.82471841372357e-54 I fail to accept the null hypothesis that republican and democrat support for the aid-to-contras bill is  different.

I never say that I "reject" the null hypothesis, I just say that I "fail to accept".

3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01

In [0]:
col = rep['crime']

np.isnan(col)

crime_no_nans = col[~np.isnan(col)]

crime_no_nans

crime_no_nans.sum()/len(crime_no_nans)

0.9813664596273292

In [0]:
rep['crime'].mean()

0.9813664596273292

In [0]:
col = dem['crime']

np.isnan(col)

crime_no_nans = col[~np.isnan(col)]

crime_no_nans

crime_no_nans.sum()/len(crime_no_nans)

0.35019455252918286

In [0]:
dem['crime'].mean()

0.35019455252918286

In [0]:
ttest_ind(rep['crime'], dem['crime'])

Ttest_indResult(statistic=nan, pvalue=nan)

In [0]:
ttest_ind(rep['crime'], dem['crime'], nan_policy='omit')

Ttest_indResult(statistic=16.342085656197696, pvalue=9.952342705606092e-47)

In [0]:
col = rep['crime']
rep_crime_no_nans = col[~np.isnan(col)]

col = dem['crime']
dem_crime_no_nans = col[~np.isnan(col)]

print(len(rep_crime_no_nans))
print(len(dem_crime_no_nans))

161
257


df = 161-1 = 160

T-statistic: 16.342085656197696

P-value: 9.952342705606092e-47

I want to reject the null hypothesis if my p-value is <.01 or if my p-value is less than (1-confidence_level)

Conclusion: due to a p-value of 9.952342705606092e-47 I fail to accept the null hypothesis that republican and democrat support for the crime bill is different.

I never say that I "reject" the null hypothesis, I just say that I "fail to accept".

4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

In [0]:
col = rep['water-project']

np.isnan(col)

water_project = col[~np.isnan(col)]

water_project

water_project.sum()/len(water_project)

0.5067567567567568

In [0]:
rep['water-project'].mean()

0.5067567567567568

In [0]:
col = dem['water-project']

np.isnan(col)

water_project = col[~np.isnan(col)]

water_project

water_project.sum()/len(south_africa)

0.821917808219178

In [0]:
dem['water-project'].mean()

0.502092050209205

In [0]:
ttest_ind(rep['water-project'], dem['water-project'])

Ttest_indResult(statistic=nan, pvalue=nan)

In [0]:
ttest_ind(rep['water-project'], dem['water-project'], nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

In [0]:
col = rep['water-project']
rep_water_project_no_nans = col[~np.isnan(col)]

col = dem['water-project']
dem_water_project_no_nans = col[~np.isnan(col)]

print(len(rep_water_project_no_nans))
print(len(dem_water_project_no_nans))

148
239


df = 148-1 = 147

T-statistic: 0.08896538137868286

P-value: 0.9291556823993485

I want to reject the null hypothesis if my p-value is <.01 or if my p-value is less than (1-confidence_level)

Conclusion: due to a p-value of 0.9291556823993485 I fail to reject the null hypothesis that republican and democrat support for the crime bill is different.

I never say that I "accept" the null hypothesis, I just say that I "fail to reject".